clients/src/main/java/org/apache/kafka/common/metrics/Metrics.java

```diff
@@ -524,6 +524,58 @@ public void addMetric(MetricName metricName, MetricValueProvider<?> metricValueP
         addMetric(metricName, null, metricValueProvider);
     }
 
+    /**
+     * Create or get an existing metric to monitor an object that implements Measurable.
+     * This metric won't be associated with any sensor. This is a way to expose existing values as metrics.
+     *
+     * This method is kept for binary compatibility purposes, it has the same behaviour as
+     * {@link #metricOrElseCreate(MetricName, MetricConfig, MetricValueProvider)}.
+     *
+     * @param metricName The name of the metric
+     * @param config The configuration to use when measuring this measurable
+     * @param measurable The measurable that will be measured by this metric
+     * @return Existing KafkaMetric if already registered or else a newly created one
+     */
+    public KafkaMetric metricOrElseCreate(MetricName metricName, MetricConfig config, Measurable measurable) {
+        return metricOrElseCreate(metricName, config, (MetricValueProvider<?>) measurable);
+    }
+
+    /**
+     * Create or get an existing metric to monitor an object that implements MetricValueProvider.
+     * This metric won't be associated with any sensor. This is a way to expose existing values as metrics.
+     * This method takes care of synchronisation while updating/accessing metrics by concurrent threads.
+     *
+     * @param metricName The name of the metric
+     * @param metricValueProvider The metric value provider associated with this metric
+     * @return Existing KafkaMetric if already registered or else a newly created one
+     */
+    public KafkaMetric metricOrElseCreate(MetricName metricName, MetricValueProvider<?> metricValueProvider) {
+        return metricOrElseCreate(metricName, null, metricValueProvider);
+    }
+
+    /**
+     * Create or get an existing metric to monitor an object that implements MetricValueProvider.
+     * This metric won't be associated with any sensor. This is a way to expose existing values as metrics.
+     * This method takes care of synchronisation while updating/accessing metrics by concurrent threads.
+     *
+     * @param metricName The name of the metric
+     * @param metricValueProvider The metric value provider associated with this metric
+     * @return Existing KafkaMetric if already registered or else a newly created one
+     */
+    public KafkaMetric metricOrElseCreate(MetricName metricName, MetricConfig config, MetricValueProvider<?> metricValueProvider) {
```

Comment:
Instead of updating ‘registerMetric’, did we consider using synchronized?

Reply:
Yeah i did consider it. What I felt was the logic remains exactly the same it's just that earlier `registerMetric` used to throw an IllegalArgumentException while with the new changes we just return. So, instead of duplicating I thought it's better to update the already existing method. Plz let me know if that makes sense.

Reply:
If we agree on the other comment I had, i.e. letting the `registerMetric` to return the existing metric, then this function can be a single op: we just blindly create the metric, and tries to call `registerMetric`, if it returns non-null value, then just return that one; otherwise return the newly created metric.

In this case we do not need the `first check, then create` two operations and worries about its atomicity.